packages/js/components/src/tree-select-control/index.js

```diff
@@ -0,0 +1,497 @@
+ * @property {string} value The value for the option
+ * @property {string} [key] Optional unique key for the Option. It will fallback to the value property if not defined
+ */
+
+/**
+ * @typedef {Object} BaseOption
+ * @property {string}   label      The label for the option
+ * @property {Option[]} [children] The children Option objects
+ *
+ * @typedef {CommonOption & BaseOption} Option
+ */
+
+/**
+ * @typedef {Object} BaseInnerOption
+ * @property {string|JSX.Element}      label          The label string or label with highlighted react element for the option.
+ * @property {InnerOption[]|undefined} children       The children options. The options are filtered if in searching.
+ * @property {boolean}                 hasChildren    Whether this option has children.
+ * @property {InnerOption[]}           leaves         All leaf options that are flattened under this option. The options are filtered if in searching.
+ * @property {boolean}                 checked        Whether this option is checked.
+ * @property {boolean}                 partialChecked Whether this option is partially checked.
+ * @property {boolean}                 expanded       Whether this option is expanded.
+ * @property {boolean}                 parent         The parent of the current option
+ * @typedef {CommonOption & BaseInnerOption} InnerOption
+ */
+
+/**
+ * Renders a component with a searchable control, tags and a tree selector.
+ *
+ * @param {Object}                     props                              Component props.
+ * @param {string}                     [props.id]                         Component id
+ * @param {string}                     [props.label]                      Label for the component
+ * @param {string | false}             [props.selectAllLabel]             Label for the Select All root element. False for disable.
+ * @param {string}                     [props.help]                       Help text under the select input.
+ * @param {string}                     [props.placeholder]                Placeholder for the search control input
+ * @param {string}                     [props.className]                  The class name for this component
+ * @param {boolean}                    [props.disabled]                   Disables the component
+ * @param {Option[]}                   [props.options]                    Options to show in the component
+ * @param {string[]}                   [props.value]                      Selected values
+ * @param {number}                     [props.maxVisibleTags]             The maximum number of tags to show. Undefined, 0 or less than 0 evaluates to "Show All".
+ * @param {Function}                   [props.onChange]                   Callback when the selector changes
+ * @param {(visible: boolean) => void} [props.onDropdownVisibilityChange] Callback when the visibility of the dropdown options is changed.
+ * @return {JSX.Element} The component
+ */
+const TreeSelectControl = ( {
```

Comment:
This is really a large component. Is it possible to split some methods into a util file?

Reply:
As far we understood it. Most of the file size is documentation and memoized functions that are depending on the component state closely. 

The rest of the functions are just handlers. Like click, tag remove, etc... some of them also depending closely to states & refs

Besides that, they are not really util function IMO (I see utils as generic functions we can use over all the files). 

My last point is that moving functions to another file will not reduce complexity and size but maybe yes readability and potentially complexity and more work in case we decide to refactor the state dependant functions to allow them to easily stay in an external file. 

If it's not a blocker I'd prefer to pass by this one. Otherwise let me know which one would be the essential functions to move.

Reply:
@puntope Thanks for the reply. 🙂 

> The rest of the functions are just handlers. Like click, tag remove, etc... some of them also depending closely to states & refs

> Besides that, they are not really util function IMO (I see utils as generic functions we can use over all the files).

I see. That makes sense to me!